Add secure functions to varrun.py

cmcantalupo opened the pull request:
- Relates to #2093
- Fixes #2094

cmcantalupo committed:
Add secure functions and doc strings

- Fixup use of umask

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

Krebrov001 committed:
Add `TestSecureFiles.py`

Change the `Makefile.mk` and `__main__.py` to include that new file.

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

Krebrov001 committed:
Added new tests related to directory creation to TestSecureFiles

- test_default_creation
- test_creation_link_not_dir
- test_creation_file_not_dir
- test_creation_bad_perms
- test_creation_bad_user_owner
- test_creation_bad_group_owner
- test_pre_exists

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

Krebrov001 committed:
Implement `secure_read_file()` and remove `secure_remove_file()`

Write implementation for `secure_read_file()` in `varrun.py` and replace
the original code in `_load_session_file()` with a call to the function.

Delete the `secure_remove_file()`, which we no longer need.

Modify the following tests in the `TestActiveSessions.py` to make them
pass with the new implementation of `secure_read_file()`.

- test_creation_bad_session_perms
- test_creation_bad_session_user_owner
- test_creation_bad_session_group_owner

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

Krebrov001 committed:
Implement `secure_make_dirs()`

Write the implementation of the `secure_make_dirs()` in the `varrun.py`
and replace the code in the `__init__()` with a call to that function.

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

bgeltz committed:
Mock secure file I/O in TestActiveSessions

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

Krebrov001 committed:
Add new tests for `secure_read_file()`

- test_read_file_not_exists
- test_read_file_is_directory
- test_read_file_is_link
- test_read_file_is_fifo
- test_read_file_bad_permissions
- test_read_file_bad_user_owner
- test_read_file_bad_group_owner
- test_read_valid_file

Modify `secure_read_file()` for better handling of symbolic links.

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

Krebrov001 committed:
Add checks to the other functions used in `mock.patch`

Previously only `sys.stderr.write` had a named object in `mock.patch`
and the calls to that function were checked.
Now we are having named objects and checked calls for all the other
functions that are used in `mock.patch`

Also updated some of the tests to use more asserts for robustness.

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

Krebrov001 committed:
Implement `secure_make_file()` and test for it

Add test_secure_make_file to the `TestSecureFiles.py` and `Makefile.mk`

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

cmcantalupo commented on service/geopmdpy_test/TestActiveSessions.py:
```diff
@@ -162,155 +162,6 @@ def create_json_file(self, directory, filename, contents, permissions=0o600):
             else:
                 json.dump(contents, file)
 
-    def test_default_creation(self):
```
Please leave these tests.  They should continue to pass without modification, and we will update them later to mock the secure_*() methods

cmcantalupo commented on service/geopmdpy_test/TestActiveSessions.py:
```diff
@@ -339,16 +192,21 @@ def test_creation_bad_session_perms(self):
         # first time in __init__() it is called with a string to determine the directory
         # second time in _load_session_file() it is called with a file descriptor to determine the file
         # based on the data type we use a different mock for the os.stat()
-        side_effect = lambda path: type(path) is int and session_2_mock or dir_mock
```
What was the reason that this unit test for the ActiveSessions class need to change with the refactor?  I don't expect TestActiveSessions.py to change at all for this PR.

Krebrov001 replied:
The implementation of `secure_read_file()` has changed, so the unit tests should be updated accordingly. We see new mocks added for `stat.S_ISREG` and `stat.S_ISLNK`. And we see an additional mock for `os.path.isdir()` which is calleld twice, once for the directory, once for the file, so we are having `os_path_isdir_effect` to distinguish between these two use cases.

bgeltz committed:
Fix failing tests in TestActiveSessions

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Preserve umask when calling make_file

- Call secure_make_file inside _update_session_file.

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

Krebrov001 committed:
Remove redundant tests from TestActiveSessions

Remove the following tests from TestActiveSessions because they either
complete duplicates or very similar to the tests in TestSecureFiles,
which cover the same run time code paths.

- test_default_creation
- test_creation_link_not_dir
- test_creation_file_not_dir
- test_creation_bad_perms
- test_creation_bad_user_owner
- test_creation_bad_group_owner
- test_creation_bad_session_perms
- test_creation_bad_session_user_owner
- test_creation_bad_session_group_owner

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

Krebrov001 committed:
Mock `secure_*()` functions for test_creation_json

- Change the `check_json_file()` which is used by test_creation_json
  in TestActiveSessions.py to use a mocks of the `secure_*()` functions,
  which are defined in varrun.py
- Because these functions are already tested and covered by TestSecureFiles.py

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

bgeltz committed:
Fix test makefile

- Test was renamed: test_update_reload ->
                    test_load_clients

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

Krebrov001 reviewed (dismissed):
I don't see anything wrong here.

bgeltz commented on service/geopmdpy/varrun.py:
```diff
@@ -52,6 +52,189 @@
+        # If it's a directory
+        else:
+            st = os.stat(path)
+            # If the permissions is not what we wanted
```
```suggestion
            # If the permissions are not what we wanted
```

bgeltz commented on service/geopmdpy/varrun.py:
```diff
@@ -52,6 +52,189 @@
+                sys.stderr.write(f'Warning: <geopm-service> {path} has wrong group owner, it will be renamed to {renamed_path}')
+                sys.stderr.write(f'Warning: <geopm-service> the wrong group owner was {group_owner}')
+                is_valid = False
+        # If one of the three above branches revealed invalid file
```
```suggestion
        # If one of the three above branches revealed an invalid file
```

bgeltz commented on service/geopmdpy/varrun.py:
```diff
@@ -52,6 +52,189 @@
+    was not secure and None is returned.
+
+    If the path points to an existing file that is determined to be
+    secure then the contents of the file is returned.
```
```suggestion
    secure then the contents of the file are returned.
```

bgeltz commented on service/geopmdpy_test/TestSecureFiles.py:
```diff
@@ -0,0 +1,591 @@
+        """
+        sess_path = f'{self._TEMP_DIR.name}/geopm-service-link'
+        os.symlink('/root', sess_path)
+        with mock.patch('os.path.islink', return_value=True) as mock_os_path_islink, \
```
It might be better here to use wraps instead of outright mocking of os.path.islink.

Krebrov001 replied:
What do you mean by wraps?

bgeltz commented on service/geopmdpy_test/TestSecureFiles.py:
```diff
@@ -0,0 +1,591 @@
+        filename = Path(sess_path)
+        filename.touch(exist_ok=True)
+
+        with mock.patch('os.path.islink', return_value=False) as mock_os_path_islink, \
```
Samee here.

bgeltz committed:
Address review feedback

- Use the 'wraps' feature of unittest.mock where
  it is applicable.
- Use autospeccing where it is applicable.

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

Krebrov001 committed:
Prevent `secure_read_file()` from opening a fifo

Previously the function `secure_read_file()` used to blindly open a fifo,
if it was provided one as the `path`, causing the process to block until
something came through the other end of the fifo. This **bug** was
causing the program to hang indefinitely.

Now I added a check so that `secure_read_file()` prints a warning
message if it determines that the passed in `path` is a fifo, and it
doesn't open it.

This change enables test_read_file_fifo in TestSecureFiles to use an
actual fifo as the input, instead of a regular file which had been
cleverly mocked to make it look like a fifo. The test is now upadted to
reflect this change.

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

bgeltz committed:
Simplify FIFO test

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

cmcantalupo commented on service/geopmdpy/varrun.py:
```diff
@@ -190,6 +191,9 @@ def secure_read_file(path):
             sys.stderr.write(f'Warning: <geopm-service> {path} is a symbolic link, it will be renamed to {renamed_path}')
             sys.stderr.write(f'Warning: <geopm-service> the symbolic link points to {os.readlink(path)}')
+        # If it is a fifo
+        elif stat.S_ISFIFO(os.stat(path).st_mode):
```
Should check for stat.S_ISREG() before opening it.  This will cover the case of a link or a FIFO.  More verbose error messaging we can check each of these individually first, but I'm also okay with a single check that it is a regular file.

bgeltz replied:
Also \n's

bgeltz committed:
Change early ISFIFO check to not ISREG

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Add \n's to warning messages

Signed-off-by: Brad Geltz <brad.geltz@intel.com>
